File: Sources/UsageEngine/Summary.swift
import Foundation

public enum Tool: String, Codable, Sendable, CaseIterable {
    case claude
    case codex
}

public struct Totals: Sendable, Equatable {
    public var cost: Double = 0
    public var inputTokens: Int = 0
    public var outputTokens: Int = 0
    public var cacheCreationTokens: Int = 0
    public var cacheReadTokens: Int = 0
    public var totalTokens: Int = 0
}

public struct DayPoint: Sendable, Equatable {
    public let date: String   // "2026-06-01"
    public let cost: Double
}

public struct ToolSlice: Sendable, Equatable {
    public let tool: Tool
    public let cost: Double
    public let totalTokens: Int
}

public struct ModelSlice: Sendable, Equatable {
    public let modelName: String
    public let tool: Tool
    public let cost: Double
    public let totalTokens: Int
}

public struct ProjectSlice: Sendable, Equatable {
    public let name: String     // display leaf, e.g. "personal"
    public let path: String     // full cwd (dedup key); "" for the Unknown bucket
    public let cost: Double
    public let totalTokens: Int
}

public struct Summary: Sendable, Equatable {
    public let today: Totals
    public let thisWeek: Totals          // rolling 7 days ending on referenceDate
    public let weekByDay: [DayPoint]     // 7 points oldest→newest, zero-filled
    public let heatmapDays: [DayPoint]   // 84 points oldest→newest, zero-filled (heatmap)
    public let byTool: [ToolSlice]       // week range
    public let byModel: [ModelSlice]     // week range, sorted by cost desc
    public let cacheSavings: Double      // week range, Claude-only, approximate
    public let monthToDate: Totals       // sum of daily rows in referenceDate's calendar month
    public let allTime: Totals           // from CcusageReport.totals
    public let avgPerDay: Double         // thisWeek.cost / 7
    public let lastWeek: Totals          // 7 days before thisWeek (offsets -13..-7)
    public let weekTrend: Double?        // (thisWeek-lastWeek)/lastWeek; nil if lastWeek.cost == 0
    public let projectedToday: Double?   // today.cost / fractionOfDayElapsed; nil if too early
    public let byProject: [ProjectSlice]   // by cwd, current data; empty if unavailable
    public let generatedAt: Date
}

File: windows/internal/engine/summary.go
package engine

import (
	"encoding/json"
	"time"
)

// Status describes how trustworthy a Summary is.
type Status string

const (
	StatusOK     Status = "ok"     // fresh data
	StatusStale  Status = "stale"  // last-good cache, fresh fetch failed
	StatusNoData Status = "noData" // nothing cached and nothing fetched
	StatusError  Status = "error"  // ccusage missing or unusable
)

// Totals is a token/cost roll-up over some window.
type Totals struct {
	Cost                float64 `json:"cost"`
	InputTokens         int64   `json:"inputTokens"`
	OutputTokens        int64   `json:"outputTokens"`
	CacheCreationTokens int64   `json:"cacheCreationTokens"`
	CacheReadTokens     int64   `json:"cacheReadTokens"`
	TotalTokens         int64   `json:"totalTokens"`
}

// add returns a+b field-wise.
func (a Totals) add(b Totals) Totals {
	return Totals{
		Cost:                a.Cost + b.Cost,
		InputTokens:         a.InputTokens + b.InputTokens,
		OutputTokens:        a.OutputTokens + b.OutputTokens,
		CacheCreationTokens: a.CacheCreationTokens + b.CacheCreationTokens,
		CacheReadTokens:     a.CacheReadTokens + b.CacheReadTokens,
		TotalTokens:         a.TotalTokens + b.TotalTokens,
	}
}

// DayPoint is one day of the sparkline or heatmap series.
type DayPoint struct {
	Date string  `json:"date"` // "2026-06-01"
	Cost float64 `json:"cost"`
}

// ToolSlice is a per-tool roll-up over the week window.
type ToolSlice struct {
	Tool   Tool    `json:"tool"`
	Cost   float64 `json:"cost"`
	Tokens int64   `json:"tokens"`
}

// ModelSlice is a per-model roll-up over the week window.
type ModelSlice struct {
	Model  string  `json:"model"`
	Tool   Tool    `json:"tool"`
	Cost   float64 `json:"cost"`
	Tokens int64   `json:"tokens"`
}

// ProjectSlice is a per-working-directory roll-up. Path is the dedup key (the full
// cwd, empty for the Unknown bucket) and stays out of the bridge JSON.
type ProjectSlice struct {
	Project string  `json:"project"`
	Path    string  `json:"-"`
	Cost    float64 `json:"cost"`
	Tokens  int64   `json:"tokens"`
}

// Summary is the single payload the UI renders. Its JSON shape is the bridge
// contract shared with ui/mock.js, so field tags here are load-bearing.
type Summary struct {
	Status      Status    `json:"status"`
	StaleReason string    `json:"staleReason"`
	FetchedAt   time.Time `json:"fetchedAt"`

	Today    Totals `json:"today"`
	Week     Totals `json:"week"`     // rolling 7 days ending today
	Month    Totals `json:"month"`    // today's calendar month to date
	AllTime  Totals `json:"allTime"`  // from Report.Totals
	LastWeek Totals `json:"lastWeek"` // the 7 days before this week

	AvgPerDay      float64  `json:"avgPerDay"`      // Week.Cost / 7
	WeekTrend      *float64 `json:"weekTrend"`      // (week-lastWeek)/lastWeek; null if lastWeek is 0
	ProjectedToday *float64 `json:"projectedToday"` // today extrapolated; null before 10% of the day

	Sparkline []DayPoint `json:"sparkline"` // 14 entries, oldest first
	Heatmap   []DayPoint `json:"heatmap"`   // 84 entries, oldest first

	ByTool       []ToolSlice    `json:"byTool"`
	ByModel      []ModelSlice   `json:"byModel"`
	ByProject    []ProjectSlice `json:"byProject"`
	CacheSavings float64        `json:"cacheSavings"`

	Wrapped Wrapped `json:"wrapped"`
}

// JSON renders the bridge payload. Nil slices are normalised to `[]` so the JS side
// never has to null-check an array.
func (s *Summary) JSON() ([]byte, error) {
	out := *s
	if out.Sparkline == nil {
		out.Sparkline = []DayPoint{}
	}
	if out.Heatmap == nil {
		out.Heatmap = []DayPoint{}
	}
	if out.ByTool == nil {
		out.ByTool = []ToolSlice{}
	}
	if out.ByModel == nil {
		out.ByModel = []ModelSlice{}
	}
	if out.ByProject == nil {
		out.ByProject = []ProjectSlice{}
	}
	if out.Wrapped.TopModels == nil {
		out.Wrapped.TopModels = []WrappedModel{}
	}
	return json.Marshal(out)
}
